### ISSUE TITLE
Warn on leading pipes in patterns

### ISSUE THREAD

**Issue by LeoniePhiline** [rust-lang/rust-clippy#10717]
### What it does

Warns on a  `|x| x` pattern, which looks like a closure.

### Lint Name

pattern_leading_pipe

### Category

suspicious

### Advantage

- Avoids confusion of "or" pattern having a leading pipe (which is discarded), with a closure, enclosing arguments in pipes.

### Drawbacks

*None*

### Example

```rust
    let (|x| x) = match x(..) {
        |_| Some(2) => |_| Some(3),
        |_| _ => unreachable!(),
    };
    assert!(matches!(x(..), |_| Some(4)));
```
(borrowed from [weird-exprs.rs](https://github.com/rust-lang/rust/blob/666fee2a5fb3a4ec75953b9e6622662ae10b9ae5/tests/ui/weird-exprs.rs#L198-L205))

Could be written as:

```rust
    let (x | x) = match x(..) {
        _ | Some(2) => |_| Some(3),
        _ | _ => unreachable!(),
    };
    assert!(matches!(x(..), _ | Some(4)));
```

**Comment by LeoniePhiline** [comment]
Covered by `rustfmt`. 

Should be closed?

**Comment by Alexendoo** [comment]
Yeah, no need for a lint if rustfmt has it covered

**Comment by llogiq** [comment]
I think we still might lint this in a `matches!(..)` macro invocation, because rustfmt doesn't cover macros. We'd still need to check if the pattern comes from another expansion, but otherwise it would be an OK lint. I think the category should still be pedantic or style rather than suspicious. Also we may try not to lint if the pattern spans multiple lines where each line starts with a pipe symbol (something I hear Haskell programmers are quite fond of).

So I think we may reopen this, but it's certainly no longer an easy issue.

**Comment by Alexendoo** [comment]
There's an open PR on the rustfmt side to fix the `matches!` case - https://github.com/rust-lang/rustfmt/pull/5554

**Comment by llogiq** [comment]
Ah, I had overlooked that. In that case, it's probably not a good idea to invest in this.
